### PR TITLE
jax/jaxlib upgrade

### PR DESCRIPTION
**Pull request by samuela**
###### Description of changes

I initially started this after forgetting about https://github.com/NixOS/nixpkgs/pull/227145, but have since gotten the build to work. For better or worse, that requires adding `bazel run` functionality to `buildBazelPackage`.

Testing on other platforms would be much appreciated!

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 237648` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>36 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.aeppl</li>
    <li>python310Packages.aesara</li>
    <li>python310Packages.arviz</li>
    <li>python310Packages.augmax</li>
    <li>python310Packages.bambi</li>
    <li>python310Packages.blackjax</li>
    <li>python310Packages.chex</li>
    <li>python310Packages.dalle-mini</li>
    <li>python310Packages.distrax</li>
    <li>python310Packages.dm-haiku</li>
    <li>python310Packages.dm-sonnet</li>
    <li>python310Packages.elegy</li>
    <li>python310Packages.flax</li>
    <li>python310Packages.jaxlib</li>
    <li>python310Packages.jaxlib-build</li>
    <li>python310Packages.jaxlibWithoutCuda</li>
    <li>python310Packages.jaxopt</li>
    <li>python310Packages.jmp</li>
    <li>python310Packages.n3fit</li>
    <li>python310Packages.numpyro</li>
    <li>python310Packages.objax</li>
    <li>python310Packages.optax</li>
    <li>python310Packages.optuna</li>
    <li>python310Packages.pymc</li>
    <li>python310Packages.pytensor</li>
    <li>python310Packages.rlax</li>
    <li>python310Packages.scikit-tda</li>
    <li>python310Packages.tensorflow-datasets</li>
    <li>python310Packages.tensorflow-probability</li>
    <li>python310Packages.treeo</li>
    <li>python310Packages.treex</li>
    <li>python310Packages.trfl</li>
    <li>python310Packages.vqgan-jax</li>
    <li>python311Packages.jaxlib</li>
    <li>python311Packages.jaxlib-build</li>
    <li>python311Packages.jaxlibWithoutCuda</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>python310Packages.textnets</li>
  </ul>
</details>
<details>
  <summary>25 packages built:</summary>
  <ul>
    <li>bazel-watcher</li>
    <li>libtensorflow</li>
    <li>openai-full</li>
    <li>python310Packages.baselines</li>
    <li>python310Packages.ctranslate2</li>
    <li>python310Packages.edward</li>
    <li>python310Packages.fastai</li>
    <li>python310Packages.gpt-2-simple</li>
    <li>python310Packages.jax</li>
    <li>python310Packages.jaxlib-bin</li>
    <li>python310Packages.k-diffusion</li>
    <li>python310Packages.langchain</li>
    <li>python310Packages.ml-dtypes</li>
    <li>python310Packages.pot</li>
    <li>python310Packages.skrl</li>
    <li>python310Packages.spacy</li>
    <li>python310Packages.spacy-loggers</li>
    <li>python310Packages.spacy-transformers</li>
    <li>python310Packages.tensorflow (python310Packages.tensorflow-build ,python310Packages.tensorflowWithoutCuda)</li>
    <li>python310Packages.textacy</li>
    <li>python310Packages.tflearn</li>
    <li>python310Packages.umap-learn</li>
    <li>python310Packages.wandb</li>
    <li>python311Packages.ml-dtypes</li>
    <li>wyoming-faster-whisper</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Hello ! Thank you for handling this !
I will include your changes to my original PR and let you know once this is done.

File: pkgs/development/python-modules/ml-dtypes/default.nix [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  version = "0.2.0";
+  # As of 2023-06-06, they have both pyproject.toml and setup.py, seems like setuptools works more easily.
+  format = "setuptools";
+
+  src = fetchFromGitHub {
+    owner = "jax-ml";
+    repo = "ml_dtypes";
+    rev = "refs/tags/v${version}";
+    # ml-dtypes has eigen pinned as a submodule. My efforts to use nixpkgs's eigen have thus far failed :/
+    fetchSubmodules = true;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/ml-dtypes/default.nix:19]: Can we link our eigen source?

File: pkgs/development/python-modules/jaxlib/default.nix [hunk 2]
```diff
@@ -168,7 +169,7 @@ let
       numpy
       openssl
       pkgs.flatbuffers
-      protobuf
+      pkgs.protobuf
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/jaxlib/default.nix:172]:
```suggestion
      protobuf
```
why is pkgs even in the inputs already?

File: pkgs/development/python-modules/jaxlib/default.nix [hunk 3]
```diff
@@ -206,7 +208,11 @@ let
       build --action_env=PYENV_ROOT
       build --python_path="${python}/bin/python"
       build --distinct_host_configuration=false
-      build --define PROTOBUF_INCLUDE_PATH="${protobuf}/include"
+      build --define PROTOBUF_INCLUDE_PATH="${pkgs.protobuf}/include"
+    '' + lib.optionalString (stdenv.targetPlatform.isx86_64 && stdenv.targetPlatform.isUnix) ''
+      build --config=avx_posix
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/jaxlib/default.nix:213]: This should be dependent on stdenv.hostPlatform.avxSupport

File: pkgs/development/python-modules/jaxlib/bin.nix [hunk 4]
```diff
@@ -123,7 +132,7 @@ buildPythonPackage rec {
     ln -s ${cudatoolkit}/bin/ptxas $out/bin/ptxas
   '';
 
-  pythonImportsCheck = [ "jaxlib" ];
+  pythonImportsCheck = jaxlib.pythonImportsCheck;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/jaxlib/bin.nix:135]:
```suggestion
  inherit (jaxlib) pythonImportsCheck;
```

File: pkgs/development/python-modules/jaxlib/bin.nix [hunk 5]
```diff
@@ -37,34 +39,40 @@ in
 
 assert cudaSupport -> lib.versionAtLeast cudatoolkit.version "11.1";
 assert cudaSupport -> lib.versionAtLeast cudnn.version "8.2";
+assert cudaSupport -> stdenv.isLinux;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/jaxlib/bin.nix:42]:
```suggestion
assert cudaSupport -> lib.versionAtLeast cudatoolkit.version "11.1" && lib.versionAtLeast cudnn.version "8.2" && stdenv.isLinux;
```

File: pkgs/build-support/build-bazel-package/default.nix [hunk 6]
```diff
@@ -262,6 +270,15 @@ stdenv.mkDerivation (fBuildAttrs // {
         targets = fBuildAttrs.bazelTargets;
       }
     }
+    ${
+      bazelCmd {
+        cmd = "run";
+        additionalFlags = fBuildAttrs.bazelRunFlags ++ ["--jobs" "$NIX_BUILD_CORES"];
+        # Bazel run only accepts a single target, but `bazelCmd` expects `targets` to be a list.
+        targets = if fBuildAttrs.bazelRunTarget == null then [] else [fBuildAttrs.bazelRunTarget];
```
Comment (SuperSandro2000) [inline pkgs/build-support/build-bazel-package/default.nix:278]:
```suggestion
        targets = lib.optional (fBuildAttrs.bazelRunTarget != null) fBuildAttrs.bazelRunTarget;
```

File: pkgs/build-support/build-bazel-package/default.nix [hunk 7]
```diff
@@ -262,6 +270,15 @@ stdenv.mkDerivation (fBuildAttrs // {
         targets = fBuildAttrs.bazelTargets;
       }
     }
+    ${
+      bazelCmd {
+        cmd = "run";
+        additionalFlags = fBuildAttrs.bazelRunFlags ++ ["--jobs" "$NIX_BUILD_CORES"];
```
Comment (SuperSandro2000) [inline pkgs/build-support/build-bazel-package/default.nix:276]:
```suggestion
        additionalFlags = fBuildAttrs.bazelRunFlags ++ [ "--jobs" "$NIX_BUILD_CORES" ];
```

File: pkgs/build-support/build-bazel-package/default.nix [hunk 8]
```diff
@@ -50,13 +53,17 @@ let
     bazelFlags = bazelFlags;
     bazelBuildFlags = bazelBuildFlags;
     bazelTestFlags = bazelTestFlags;
+    bazelRunFlags = bazelRunFlags;
+    runTargetFlags = runTargetFlags;
     bazelFetchFlags = bazelFetchFlags;
+    bazelTargets = bazelTargets;
     bazelTestTargets = bazelTestTargets;
+    bazelRunTarget = bazelRunTarget;
     dontAddBazelOpts = dontAddBazelOpts;
```
Comment (SuperSandro2000) [inline pkgs/build-support/build-bazel-package/default.nix:62]: Why not do inherit here?

File: pkgs/development/python-modules/ml-dtypes/default.nix (hunk 1 above)
Comment (samuela) [reply pkgs/development/python-modules/ml-dtypes/default.nix:19]: I made an attempt at this but was not able to manage it. Happy to update if someone can figure out how to make it happen!

File: pkgs/development/python-modules/jaxlib/default.nix (hunk 2 above)
Comment (samuela) [reply pkgs/development/python-modules/jaxlib/default.nix:172]: good question! when attribute names in `pkgs` and `python3Packages` conflict, argumes evaluate to the `python3Packages` one. When we actually require the `pkgs` one, we have to specify as such. it looks odd at first but fwiw I am not introducing this pattern in this PR

File: pkgs/development/python-modules/jaxlib/default.nix (hunk 3 above)
Comment (samuela) [reply pkgs/development/python-modules/jaxlib/default.nix:213]: ah didn't know that existed! would that be 
```
    '' + lib.optionalString (stdenv.targetPlatform.avxSupport && stdenv.targetPlatform.isUnix) ''
```
then?

File: pkgs/build-support/build-bazel-package/default.nix (hunk 6 above)
Comment (samuela) [reply pkgs/build-support/build-bazel-package/default.nix:278]: ah good catch!

File: pkgs/build-support/build-bazel-package/default.nix (hunk 7 above)
Comment (samuela) [reply pkgs/build-support/build-bazel-package/default.nix:276]: I agree that the with-spaces formatting is preferable, but in this case I was intentionally following the formatting as is used in the rest of this file, esp in the lines immediately preceding this block

I'd be happy proceeding with either version

File: pkgs/build-support/build-bazel-package/default.nix (hunk 8 above)
Comment (samuela) [reply pkgs/build-support/build-bazel-package/default.nix:62]: in this case I was simply following the source as it's currently written, aiming to minimize git blame noise

File: pkgs/build-support/build-bazel-package/default.nix (hunk 7 above)
Comment (samuela) [reply pkgs/build-support/build-bazel-package/default.nix]: done in https://github.com/NixOS/nixpkgs/pull/237648/commits/9114a2593e9f7a5f215a8692e7f73df903b0e133

File: pkgs/build-support/build-bazel-package/default.nix (hunk 6 above)
Comment (samuela) [reply pkgs/build-support/build-bazel-package/default.nix]: done in https://github.com/NixOS/nixpkgs/pull/237648/commits/9114a2593e9f7a5f215a8692e7f73df903b0e133

File: pkgs/development/python-modules/jaxlib/default.nix (hunk 3 above)
Comment (samuela) [reply pkgs/development/python-modules/jaxlib/default.nix:213]: done in https://github.com/NixOS/nixpkgs/pull/237648/commits/d2f59991294188e83b9a61c92045fbbea364f37b

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 237648` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>18 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.dalle-mini</li>
    <li>python310Packages.dalle-mini.dist</li>
    <li>python310Packages.distrax</li>
    <li>python310Packages.distrax.dist</li>
    <li>python310Packages.elegy</li>
    <li>python310Packages.elegy.dist</li>
    <li>python310Packages.flax</li>
    <li>python310Packages.flax.dist</li>
    <li>python310Packages.mask-rcnn</li>
    <li>python310Packages.mask-rcnn.dist</li>
    <li>python310Packages.optuna</li>
    <li>python310Packages.optuna.dist</li>
    <li>python310Packages.rlax</li>
    <li>python310Packages.rlax.dist</li>
    <li>python310Packages.treex</li>
    <li>python310Packages.treex.dist</li>
    <li>python310Packages.vqgan-jax</li>
    <li>python310Packages.vqgan-jax.dist</li>
  </ul>
</details>
<details>
  <summary>43 packages failed to build:</summary>
  <ul>
    <li>openai-full</li>
    <li>openai-full.dist</li>
    <li>python310Packages.arviz</li>
    <li>python310Packages.arviz.dist</li>
    <li>python310Packages.bambi</li>
    <li>python310Packages.bambi.dist</li>
    <li>python310Packages.dm-sonnet</li>
    <li>python310Packages.dm-sonnet.dist</li>
    <li>python310Packages.fastai</li>
    <li>python310Packages.fastai.dist</li>
    <li>python310Packages.k-diffusion</li>
    <li>python310Packages.k-diffusion.dist</li>
    <li>python310Packages.langchain</li>
    <li>python310Packages.langchain.dist</li>
    <li>python310Packages.numpyro</li>
    <li>python310Packages.numpyro.dist</li>
    <li>python310Packages.pymc</li>
    <li>python310Packages.pymc.dist</li>
    <li>python310Packages.scikit-tda</li>
    <li>python310Packages.scikit-tda.dist</li>
    <li>python310Packages.skrl</li>
    <li>python310Packages.skrl.dist</li>
    <li>python310Packages.spacy</li>
    <li>python310Packages.spacy-loggers</li>
    <li>python310Packages.spacy-loggers.dist</li>
    <li>python310Packages.spacy-transformers</li>
    <li>python310Packages.spacy-transformers.dist</li>
    <li>python310Packages.spacy.dist</li>
    <li>python310Packages.tensorflow-datasets</li>
    <li>python310Packages.tensorflow-datasets.dist</li>
    <li>python310Packages.textacy</li>
    <li>python310Packages.textacy.dist</li>
    <li>python310Packages.textnets</li>
    <li>python310Packages.textnets.dist</li>
    <li>python310Packages.wandb</li>
    <li>python310Packages.wandb.dist</li>
    <li>python311Packages.dm-haiku</li>
    <li>python311Packages.dm-haiku.dist</li>
    <li>python311Packages.dm-haiku.testsout</li>
    <li>python311Packages.jmp</li>
    <li>python311Packages.jmp.dist</li>
    <li>tts</li>
    <li>tts.dist</li>
  </ul>
</details>
<details>
  <summary>94 packages built:</summary>
  <ul>
    <li>bazel-watcher</li>
    <li>envoy</li>
    <li>frigate</li>
    <li>gpt2tc</li>
    <li>libtensorflow (libtensorflow.python)</li>
    <li>pomerium</li>
    <li>python310Packages.aeppl</li>
    <li>python310Packages.aeppl.dist</li>
    <li>python310Packages.aesara</li>
    <li>python310Packages.aesara.dist</li>
    <li>python310Packages.augmax</li>
    <li>python310Packages.augmax.dist</li>
    <li>python310Packages.baselines</li>
    <li>python310Packages.baselines.dist</li>
    <li>python310Packages.blackjax</li>
    <li>python310Packages.blackjax.dist</li>
    <li>python310Packages.chex</li>
    <li>python310Packages.chex.dist</li>
    <li>python310Packages.ctranslate2</li>
    <li>python310Packages.ctranslate2.dist</li>
    <li>python310Packages.dm-haiku</li>
    <li>python310Packages.dm-haiku.dist</li>
    <li>python310Packages.dm-haiku.testsout</li>
    <li>python310Packages.edward</li>
    <li>python310Packages.edward.dist</li>
    <li>python310Packages.faster-whisper</li>
    <li>python310Packages.faster-whisper.dist</li>
    <li>python310Packages.gpt-2-simple</li>
    <li>python310Packages.gpt-2-simple.dist</li>
    <li>python310Packages.jax</li>
    <li>python310Packages.jax.dist</li>
    <li>python310Packages.jaxlib (python310Packages.jaxlib-build ,python310Packages.jaxlibWithoutCuda)</li>
    <li>python310Packages.jaxlib-bin</li>
    <li>python310Packages.jaxlib-bin.dist</li>
    <li>python310Packages.jaxlib.dist (python310Packages.jaxlib-build.dist ,python310Packages.jaxlibWithoutCuda.dist)</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.jaxlibWithCuda.dist</li>
    <li>python310Packages.jaxopt</li>
    <li>python310Packages.jaxopt.dist</li>
    <li>python310Packages.jmp</li>
    <li>python310Packages.jmp.dist</li>
    <li>python310Packages.ml-dtypes</li>
    <li>python310Packages.ml-dtypes.dist</li>
    <li>python310Packages.n3fit</li>
    <li>python310Packages.n3fit.dist</li>
    <li>python310Packages.objax</li>
    <li>python310Packages.objax.dist</li>
    <li>python310Packages.optax</li>
    <li>python310Packages.optax.dist</li>
    <li>python310Packages.optax.testsout</li>
    <li>python310Packages.pot</li>
    <li>python310Packages.pot.dist</li>
    <li>python310Packages.pytensor</li>
    <li>python310Packages.pytensor.dist</li>
    <li>python310Packages.tensorflow (python310Packages.tensorflow-build ,python310Packages.tensorflowWithoutCuda)</li>
    <li>python310Packages.tensorflow-bin</li>
    <li>python310Packages.tensorflow-bin.dist</li>
    <li>python310Packages.tensorflow.dist (python310Packages.tensorflow-build.dist ,python310Packages.tensorflowWithoutCuda.dist)</li>
    <li>python310Packages.tensorflow-probability</li>
    <li>python310Packages.tensorflow-probability.dist</li>
    <li>python310Packages.tensorflowWithCuda</li>
    <li>python310Packages.tensorflowWithCuda.dist</li>
    <li>python310Packages.tflearn</li>
    <li>python310Packages.tflearn.dist</li>
    <li>python310Packages.treeo</li>
    <li>python310Packages.treeo.dist</li>
    <li>python310Packages.trfl</li>
    <li>python310Packages.trfl.dist</li>
    <li>python310Packages.umap-learn</li>
    <li>python310Packages.umap-learn.dist</li>
    <li>python311Packages.augmax</li>
    <li>python311Packages.augmax.dist</li>
    <li>python311Packages.blackjax</li>
    <li>python311Packages.blackjax.dist</li>
    <li>python311Packages.chex</li>
    <li>python311Packages.chex.dist</li>
    <li>python311Packages.jax</li>
    <li>python311Packages.jax.dist</li>
    <li>python311Packages.jaxlib (python311Packages.jaxlib-build ,python311Packages.jaxlibWithoutCuda)</li>
    <li>python311Packages.jaxlib.dist (python311Packages.jaxlib-build.dist ,python311Packages.jaxlibWithoutCuda.dist)</li>
    <li>python311Packages.jaxlibWithCuda</li>
    <li>python311Packages.jaxlibWithCuda.dist</li>
    <li>python311Packages.jaxopt</li>
    <li>python311Packages.jaxopt.dist</li>
    <li>python311Packages.ml-dtypes</li>
    <li>python311Packages.ml-dtypes.dist</li>
    <li>python311Packages.optax</li>
    <li>python311Packages.optax.dist</li>
    <li>python311Packages.optax.testsout</li>
    <li>python311Packages.treeo</li>
    <li>python311Packages.treeo.dist</li>
    <li>verible</li>
    <li>wyoming-faster-whisper</li>
    <li>wyoming-faster-whisper.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
As the update is currently being worked in in #227145, we might close this PR right ?

**Pull request closed without merging**